scripts/playbooks/roles/preflight_checks/tasks/check_plugins.yaml

```diff
@@ -0,0 +1,40 @@
+---
+- set_fact:
+    base_dir: "{{ playbook_dir | dirname}}"
+
+# Create config file for jenkins plugin install
+- name: Generate jenkins plugin install template
+  template:
+    src: "jpi-config.ini.j2"
+    dest: "{{base_dir}}/jpi-config.ini"
+
+- name: Get a list of installed plugins
+  shell: "docker run --privileged --rm -v {{ base_dir }}:{{ base_dir }} quay.io/feedhenry/jenkins-plugin-install jpi list --conf {{ base_dir }}/jpi-config.ini"
+  register: get_installed_plugins
+
+- set_fact:
+    installed_plugins: {}
+    plugins_to_install: {}
+    incompatible_plugins: {}
+
+- block:
+  - name: Parse list of installed plugins
+    set_fact:
+      installed_plugins: "{{ installed_plugins | combine({item.split(':')[0]: item.split(':')[1]}) }}"
+    with_items: "{{ get_installed_plugins.stdout_lines}}"
+  - name: Parse list of plugins to install
+    set_fact: 
+      plugins_to_install: "{{ plugins_to_install | combine({item.split(':')[0]: item.split(':')[1]}) }}"
+    with_items: "{{ lookup('file', '{{ base_dir }}/plugins.txt').splitlines() }}"
+  - name: Get list of incompatible plugins
+    set_fact:
+      incompatible_plugins: "{{ incompatible_plugins | combine({plugin.key: {'current_version': installed_plugins[plugin.key], 'target_version': plugin.value }}) }}"
+    when: installed_plugins[plugin.key] is defined and installed_plugins[plugin.key] != "{{ plugin.value | replace(' ', '') }}"
+    with_dict: "{{ plugins_to_install }}"
+    loop_control:
+      loop_var: "plugin"
+  when: get_installed_plugins.stdout_lines | length > 0
+
+- fail:
+    msg: "Found the following incompatible plugins, please update these to the target version. {{ incompatible_plugins }}"
+  when: incompatible_plugins != {}
```

Comment:
So this will just fail and tell the person they need to update these plugins manually themselves if they want to continue?

Reply:
Did a test and that's what is happening anyway:

```TASK [preflight_checks : fail] ***********************************************************************************************************************************************************************************************************************************************************
fatal: [127.0.0.1]: FAILED! => {"changed": false, "msg": "Found the following incompatible plugins, please update these to the target version. {'github': {'current_version': '1.29.3', 'target_version': '1.29.4'}}"}

PLAY RECAP *******************************************************************************************************************************************************************************************************************************************************************************
127.0.0.1                  : ok=13   changed=1    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0   ```

What happens if we try and install the version of the plugin we are specifying? I'm just wondering if we should have it fail here or not.

Reply:
@mikenairn yes. They either need to manually update these or delete them so that they get re-installed with the correct target version. Our cli for installing the plugins, skips the installation of ones that are already installed.